File: src/usage/http.ts
/**
 * Shared HTTP for usage fetchers. One timeout, one retry policy.
 * Port of `legacy-rust/src/usage/http.rs`:
 *   - `httpJson` is the simple wrapper used by fetchers that don't need retry.
 *   - `sendRetry` re-issues on 429 / 5xx with exponential backoff 500ms*2^n
 *     capped at 5 attempts (per the Phase 4 spec).
 */

const DEFAULT_TIMEOUT_MS = 20_000;
const MAX_ATTEMPTS = 5;
const BASE_BACKOFF_MS = 500;

async function withTimeout(
  init: RequestInit | undefined,
  timeoutMs: number,
): Promise<RequestInit> {
  const signal = (init?.signal ?? undefined) as AbortSignal | undefined;
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(new Error("timeout")), timeoutMs);
  if (signal) {
    if (signal.aborted) ctrl.abort(signal.reason);
    else signal.addEventListener("abort", () => ctrl.abort(signal.reason), { once: true });
  }
  return { ...(init ?? {}), signal: ctrl.signal };
}

/** GET/POST/etc. with a 20s default timeout. Throws on non-2xx with status+body. */
export async function httpJson(
  url: string,
  init?: RequestInit,
  timeoutMs: number = DEFAULT_TIMEOUT_MS,
): Promise<unknown> {
  const merged = await withTimeout(init, timeoutMs);
  const resp = await fetch(url, merged);
  if (!resp.ok) {
    const body = await resp.text().catch(() => "");
    throw new Error(`HTTP ${resp.status} ${resp.statusText} from ${url}: ${body}`);
  }
  return resp.json();
}

/**
 * `sendRetry` re-issues the request on 429 / 5xx (and network errors) with
 * exponential backoff (500ms * 2^attempt, attempt capped at 5). Caps at 5
 * total attempts before returning the last transient response or throwing.
 */
export async function sendRetry(
  url: string,
  init?: RequestInit,
  timeoutMs: number = DEFAULT_TIMEOUT_MS,
): Promise<Response> {
  let lastErr: unknown;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    try {
      const merged = await withTimeout(init, timeoutMs);
      const resp = await fetch(url, merged);
      if ((resp.status === 429 || resp.status >= 500) && attempt + 1 < MAX_ATTEMPTS) {
        const { promise, resolve } = Promise.withResolvers<void>();
        setTimeout(resolve, BASE_BACKOFF_MS * (1 << Math.min(attempt, 5)));
        await promise;
        continue;
      }
      return resp;
    } catch (e) {
      lastErr = e;
      if (attempt + 1 < MAX_ATTEMPTS) {
        const { promise, resolve } = Promise.withResolvers<void>();
        setTimeout(resolve, BASE_BACKOFF_MS * (1 << Math.min(attempt, 5)));
        await promise;
        continue;
      }
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(String(lastErr));
}

File: src/usage/http.rs
//! Shared HTTP for usage fetchers: one reqwest client + a small retry wrapper
//! that re-issues on 429 / 5xx with exponential backoff.
use std::time::Duration;

pub fn client() -> reqwest::Client {
    reqwest::Client::builder()
        .timeout(Duration::from_secs(20))
        .build()
        .unwrap_or_else(|_| reqwest::Client::new())
}

/// Send the request built by `build`, retrying up to `attempts` times on
/// transient failures (network error, HTTP 429, or 5xx). `build` is called
/// fresh per attempt so no request-body cloning is needed.
pub async fn send_retry(
    build: impl Fn() -> reqwest::RequestBuilder,
    attempts: usize,
) -> reqwest::Result<reqwest::Response> {
    let attempts = attempts.max(1);
    let mut last: Option<reqwest::Error> = None;
    for attempt in 0..attempts {
        match build().send().await {
            Ok(resp) => {
                let s = resp.status();
                let transient = s.as_u16() == 429 || s.is_server_error();
                if transient && attempt + 1 < attempts {
                    backoff(attempt).await;
                    continue;
                }
                return Ok(resp);
            }
            Err(e) => {
                last = Some(e);
                if attempt + 1 < attempts {
                    backoff(attempt).await;
                    continue;
                }
            }
        }
    }
    Err(last.expect("loop ran at least once and either returned or set last"))
}

async fn backoff(attempt: usize) {
    let ms = 500u64.saturating_mul(1u64 << attempt.min(5));
    tokio::time::sleep(Duration::from_millis(ms)).await;
}
